src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -494,6 +503,121 @@ private static SetTargetResult TrySetTargetField(ref GenerationContext context,
+            {
+                Parameter parameter = reader.GetParameter(parameterHandle);
+                CustomAttribute? unsafeAccessorTypeAttribute = null;
+                foreach (CustomAttributeHandle customAttributeHandle in parameter.GetCustomAttributes())
+                {
+                    reader.GetAttributeNamespaceAndName(customAttributeHandle, out StringHandle namespaceName, out StringHandle name);
+                    if (reader.StringComparer.Equals(namespaceName, "System.Runtime.CompilerServices")
+                        && reader.StringComparer.Equals(name, "UnsafeAccessorTypeAttribute"))
+                    {
+                        unsafeAccessorTypeAttribute = reader.GetCustomAttribute(customAttributeHandle);
+                        break;
+                    }
+                }
+
+                if (unsafeAccessorTypeAttribute is null)
+                {
+                    continue;
+                }
+
+                if (parameter.SequenceNumber > originalSignature.Length)
+                {
+                    return SetTargetResult.Invalid;
+                }
+
+                bool isReturnValue = parameter.SequenceNumber == 0;
+
+                TypeDesc initialType = isReturnValue ? originalSignature.ReturnType : originalSignature[parameter.SequenceNumber - 1];
+
+                bool initialTypeIsByRef = initialType.IsByRef;
+
+                if (initialTypeIsByRef)
+                {
+                    // We need to strip the byref off the type to get the
+                    // underlying type for the parameter.
+                    initialType = ((ParameterizedType)initialType).ParameterType;
+                }
+
+                if (initialType != method.Context.GetWellKnownType(WellKnownType.Object))
+                {
+                    // Illegal argument type for UnsafeAccessorTypeAttribute target
+                    return SetTargetResult.Invalid;
+                }
+
+                CustomAttributeValue<TypeDesc> decoded = unsafeAccessorTypeAttribute.Value.DecodeValue(
+                    new CustomAttributeTypeProvider(method.Module));
+
+                if (decoded.FixedArguments[0].Value is not string replacementTypeName)
+                {
+                    return SetTargetResult.Invalid;
+                }
+
+                TypeDesc replacementType = method.Module.GetTypeByCustomAttributeTypeName(
+                    replacementTypeName,
+                    throwIfNotFound: false,
+                    canonGenericResolver: (module, name) =>
+                    {
+                        if (!name.StartsWith('!'))
+                        {
+                            return null;
+                        }
+
+                        bool isMethodParameter = name.StartsWith("!!", StringComparison.Ordinal);
+
+                        if (!int.TryParse(name.AsSpan(isMethodParameter ? 2 : 1), out int index))
```

Comment:
I think we should be letting exceptions flow here. I wouldn't do a try parse and instead just assume the syntax is correct.

Reply:
This is in the ILC compiler, so letting this flow would crash the compiler. Right now (with the current code in this PR) a failure to parse will result in a (correctly handled) `TypeSystemException.TypeLoadException`

Reply:
Does this represent a compilation error a run-time error?

Reply:
Throwing a `TypeSystemException.TypeLoadException` represents a run-time error (the IL of the stub becomes just a `throw System.TypeLoadException` stub). It's the same experience as when the type being searched for is not found.

Reply:
This should really be a compile time error. We know the format is wrong and we should be able to detect if the index is out of bounds. Is that possible?

Reply:
Actually... if this is a build failure then tests get weird. Ugh... fine. Let's keep this as is. Thanks.